### PR TITLE
Altered bandwidth usage in Bandwidth test

### PR DESCRIPTION
**Pull request by m-samiec**
Altered the Bandwidth component to update the replicated state every frame instead of waiting a random period (up to 1 sec) before altering the replicated state. This behaviour seems more natural, and plays nicer with improving results when using QBI-F.

Also added a debug tool that outputs the grouping of actors into each QBI-F bucket. This was used for debugging purposes when investigating this issue, and seems useful, although I've default disabled it as it's not optimised.

Latest replication component BP - https://blueprintue.com/blueprint/vty3p9he/

NFR runs - https://buildkite.com/improbable/unrealgdk-nfr/builds/5848
GDK with frequency - 914668 & 912261
Native - 990776 & 976287
GDK without frequency - 1010384

### CODE REVIEW

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp [hunk 1]
```diff
@@ -1054,6 +1063,83 @@ void ABenchmarkGymGameModeBase::CheckVelocityForPlayerMovement()
...
+
+				int TotalCount = 0;
+				FString CountsAsString;
+				for (int Count : CountPerBucket)
+				{
+					CountsAsString += FString::Format(TEXT(" {0}"), { Count });
+					TotalCount += Count;
+				}
+
+				UE_LOG(LogBenchmarkGymGameModeBase, Warning, TEXT("Density: %s (%d)"), *CountsAsString, TotalCount);
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:1128]:
```suggestion
				UE_LOG(LogBenchmarkGymGameModeBase, Log, TEXT("Density: %s (%d)"), *CountsAsString, TotalCount);
```

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp [hunk 2]
```diff
@@ -1054,6 +1063,83 @@ void ABenchmarkGymGameModeBase::CheckVelocityForPlayerMovement()
...
+			}
+
+			int TotalCount = 0;
+			FString CountsAsString;
+			for (int Count : TotalCountPerBucket)
+			{
+				CountsAsString += FString::Format(TEXT(" {0}"), { Count });
+				TotalCount += Count;
+			}
+			UE_LOG(LogBenchmarkGymGameModeBase, Warning, TEXT("Density for all: %s (%d)"), *CountsAsString, TotalCount);
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:1138]:
```suggestion
			UE_LOG(LogBenchmarkGymGameModeBase, Log, TEXT("Density for all: %s (%d)"), *CountsAsString, TotalCount);
```

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp [hunk 3]
```diff
@@ -1054,6 +1063,83 @@ void ABenchmarkGymGameModeBase::CheckVelocityForPlayerMovement()
...
+			UGameplayStatics::GetAllActorsOfClass(GameMode->GetWorld(), GameMode->SimulatedPlayerControllerClass, PlayerControllers);
+			UGameplayStatics::GetAllActorsOfClass(GameMode->GetWorld(), GameMode->SimulatedPawnClass, PlayerCharacters);
+			UGameplayStatics::GetAllActorsOfClass(GameMode->GetWorld(), GameMode->NPCClass, NPCs);
+			AllCharacters = PlayerCharacters;
+			AllCharacters.Append(NPCs);
+
+			const USpatialGDKSettings* SpatialGDKSettings = GetDefault<USpatialGDKSettings>();
+			TArray<float> NCDDistanceRatios;
+			NCDDistanceRatios.Push(SpatialGDKSettings->FullFrequencyNetCullDistanceRatio);
+			for (auto i : SpatialGDKSettings->InterestRangeFrequencyPairs)
```
Comment (MatthewSandfordImprobable) [inline Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:1088]: Const auto&

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp [hunk 4]
```diff
@@ -1054,6 +1063,83 @@ void ABenchmarkGymGameModeBase::CheckVelocityForPlayerMovement()
...
+					CountsAsString += FString::Format(TEXT(" {0}"), { Count });
+					TotalCount += Count;
+				}
+
+				UE_LOG(LogBenchmarkGymGameModeBase, Log, TEXT("Density: %s (%d)"), *CountsAsString, TotalCount);
+			}
+
+			int TotalCount = 0;
+			FString CountsAsString;
+			for (int Count : TotalCountPerBucket)
```
Comment (MatthewSandfordImprobable) [inline Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:1133]: Possibly could use a lambda for printing out this data as we kinda use do the same logic twice except for slight string change... but just if you like...

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp [hunk 5]
```diff
@@ -1054,6 +1063,83 @@ void ABenchmarkGymGameModeBase::CheckVelocityForPlayerMovement()
...
+				NCDDistanceRatios.Push(i.DistanceRatio);
+			}
+			NCDDistanceRatios.Push(MAX_FLT);
+			NCDDistanceRatios.Sort();
+
+			TArray<int> TotalCountPerBucket;
+			TArray<int> CountPerBucket;
+			TotalCountPerBucket.Init(0, NCDDistanceRatios.Num());
+
+			for (AActor* PlayerController : PlayerControllers)
```
Comment (MatthewSandfordImprobable) [inline Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:1099]: Probably can be const?

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp [hunk 6]
```diff
@@ -1054,6 +1063,83 @@ void ABenchmarkGymGameModeBase::CheckVelocityForPlayerMovement()
...
+			{
+				CountPerBucket.Init(0, NCDDistanceRatios.Num());
+
+				FVector Pos = SpatialDriver->GetActorChannelByEntityId(SpatialDriver->GetActorEntityId(*PlayerController))->GetLastUpdatedSpatialPosition();
+				for (AActor* Character : AllCharacters)
```
Comment (MatthewSandfordImprobable) [inline Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:1104]: Probably can be const?

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp [hunk 7]
```diff
@@ -21,6 +22,7 @@
 #include "TimerManager.h"
 #include "UserExperienceComponent.h"
 #include "GameFramework/CharacterMovementComponent.h"
+#include "Utils/SpatialActorUtils.h"
```
Comment (MatthewSandfordImprobable) [inline Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:25]: Are you using this?

File: .gitignore [hunk 8]
```diff
@@ -23,6 +23,7 @@ build/
 Game/Plugins/UnrealGDK/
 spatial/schema
 spatial/snapshots
+spatial/tools
```
Comment (MatthewSandfordImprobable) [inline .gitignore:26]: Driveby?

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp [hunk 9]
```diff
@@ -1054,6 +1063,83 @@ void ABenchmarkGymGameModeBase::CheckVelocityForPlayerMovement()
...
+			AllCharacters.Append(NPCs);
+
+			const USpatialGDKSettings* SpatialGDKSettings = GetDefault<USpatialGDKSettings>();
+			TArray<float> NCDDistanceRatios;
+			NCDDistanceRatios.Push(SpatialGDKSettings->FullFrequencyNetCullDistanceRatio);
+			for (auto i : SpatialGDKSettings->InterestRangeFrequencyPairs)
+			{
+				NCDDistanceRatios.Push(i.DistanceRatio);
+			}
+			NCDDistanceRatios.Push(MAX_FLT);
```
Comment (improbable-danny) [inline Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:1092]: Does this break the full frequency bucket into two?

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp [hunk 10]
```diff
@@ -1054,6 +1063,83 @@ void ABenchmarkGymGameModeBase::CheckVelocityForPlayerMovement()
...
+							break;
+						}
+					}
+				}
+
+				int TotalCount = 0;
+				FString CountsAsString;
+				for (int Count : CountPerBucket)
+				{
+					CountsAsString += FString::Format(TEXT(" {0}"), { Count });
```
Comment (improbable-danny) [inline Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:1124]: Would it be useful to also output the ratio/distances for the buckets?

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp (hunk 5 above)
Comment (improbable-danny) [inline Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:1099]: Can NCD apply to more than one controller?

Comment (improbable-danny) [reply Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:1099]: I guess technically yes because each will have their own set of queries?

**Review by improbable-danny — APPROVED** [review]

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp (hunk 5 above)
Comment (improbable-danny) [reply Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:1099]: Oh duh, this only runs on the server :)

File: .gitignore (hunk 8 above)
Comment (m-samiec) [reply .gitignore:26]: Yeah

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp (hunk 7 above)
Comment (m-samiec) [reply Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:25]: Ahh, not anymore - will remove

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp (hunk 9 above)
Comment (m-samiec) [reply Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:1092]: Will add a comment - this is to catch actors that are outside a client's interest range

File: Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp (hunk 10 above)
Comment (m-samiec) [reply Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp:1124]: Good idea

**Review by MatthewSandfordImprobable — APPROVED** [review]
